File: static/games/crazyeights.js
/*
==================================
        SERVER -> CLIENT
==================================

* crazyeights_relay_player_info
    data: { 
        "id": string, 
        "username": string, 
        "hand": list[str], 
        "state": string, 
        "game_score": int
    }

* crazyeights_relay_board_info
    data: { 
        "top_card": string, 
        "current_suit": string,
        "current_value": string
    }

* crazyeights_message
    data: { 
        "msg": string
    }

* crazyeights_prompt_suit_choice
    data: { 
        "id": string
    }

* crazyeights_room_size
    data: {
        "room_size": int
    }

* crazyeights_game_over
    data: {
        "winner": string,
        "scores": list[int]
    }

==================================
        CLIENT -> SERVER
==================================

* crazyeights_player_join
  data: {}

* crazyeights_play_card
    data: { 
        "card_code": string
    }

* crazyeights_choose_suit
    data: { 
        "suit": string (e.g., "H") 
    }
*/


console.log("crazyeights.js is running...")

const statusText = document.getElementById("status-text");
const myHand = document.getElementById("hand-bottom");
const discardPile = document.getElementById("discard-pile");
const suitIcon = document.getElementById("current-suit-icon");
const suitPicker = document.getElementById("suit-picker");
const suitButtons = document.querySelectorAll(".suit-button");
const leaveButton = document.getElementById("leave-button");

let myId = null;
let opponentSeats = {}; 
let availableSeats = [];

const socket = io();

socket.on('connect', function () {
    myId = socket.id;
    statusText.innerText = "Connected!";
    socket.emit('crazyeights_player_join'); 
});

socket.on('crazyeights_room_size', function (data) {
    const roomSize = data.room_size;

    if (roomSize === 2) {
        availableSeats = ["top"]; 
        document.getElementById("area-top").classList.remove("hidden");
    } 
    else if (roomSize === 3) {
        availableSeats = ["left", "top"]; 
        document.getElementById("area-left").classList.remove("hidden");
        document.getElementById("area-top").classList.remove("hidden");
    } 
    else if (roomSize >= 4) {
        availableSeats = ["left", "top", "right"];
        document.getElementById("area-left").classList.remove("hidden");
        document.getElementById("area-top").classList.remove("hidden");
        document.getElementById("area-right").classList.remove("hidden");
    }
});

socket.on('crazyeights_relay_player_info', function (data) {

    const gameEndPopup = document.getElementById("game-end-popup");
    if (gameEndPopup) gameEndPopup.classList.add("hidden");

    if (data.id === myId) {
        if (data.hand) {
            drawMyHand(data.hand);
        }
        if (data.game_score !== undefined) {
            document.getElementById("score-bottom").innerText = `${data.game_score} pts`;
        }
    } 
    else {
        if (!opponentSeats[data.username]) {
            opponentSeats[data.username] = availableSeats.shift(); 
        }

        const position = opponentSeats[data.username];

        const nameElement = document.getElementById(`name-${position}`);
        if (data.username !== undefined) {
            nameElement.innerText = data.username;
        }

        const scoreElement = document.getElementById(`score-${position}`);
        if (data.game_score !== undefined) {
            scoreElement.innerText = `${data.game_score} pts`;
        }

        if (data.hand) {
            drawOpponentHand(position, data.hand.length);
        }
    }

    if (data.state !== undefined) {
        let areaElement = null;

        if (data.id === myId) {
            areaElement = document.getElementById("area-bottom");
        } 
        else {
            areaElement = document.getElementById(`area-${opponentSeats[data.username]}`);
        }

        if (data.state === "playing") {

            document.querySelectorAll('.player-area').forEach(element => {
                element.classList.remove('active-turn');
            });
            
            areaElement.classList.add('active-turn');
        } 
        else {
            areaElement.classList.remove('active-turn');
        }
    }
});

socket.on('crazyeights_relay_board_info', function (data) {
    discardPile.innerHTML = `<img src="https://deckofcardsapi.com/static/img/${data.top_card}.png" class="card-image" alt="${data.top_card}">`;
    updateSuitIcon(data.current_suit);
});

socket.on('crazyeights_message', function (data) {
    console.log("Game Server:", data.msg);
    statusText.innerText = data.msg;
});

socket.on('crazyeights_prompt_suit_choice', function (data) {
    if (data.id === myId) {
        suitPicker.classList.remove("hidden");
    }
});

socket.on('crazyeights_game_over', function (data) {
    document.getElementById("game-winner-text").innerText = `🏆 ${data.winner} is the Winner! 🏆`;

    let sortedScores = data.scores.sort((a, b) => a.score - b.score);

    const scoreList = document.getElementById("game-scores-list");
    scoreList.innerHTML = ""; 

    sortedScores.forEach(player => {
        let li = document.createElement("li");
        
        if (player.username === data.winner) {
            li.style.color = "#ffd700";
            li.innerHTML = `<strong>${player.username}: ${player.score} pts 👑</strong>`;
        } else {
            li.innerText = `${player.username}: ${player.score} pts`;
        }
        
        scoreList.appendChild(li);
    });

    document.getElementById("game-end-popup").classList.remove("hidden");
});

// ------------------------------- //

suitButtons.forEach(btn => {
    btn.addEventListener("click", function (event) {
        const chosenSuit = event.target.dataset.suit; 
        socket.emit("crazyeights_choose_suit", {suit: chosenSuit});
        suitPicker.classList.add("hidden"); 
    });
});

leaveButton.addEventListener("click", async function (event) {
    socket.emit("blackjack_leave_room");
    await fetch("/leave_room", {
        method:"POST", 
        credentials: "same-origin"
    });
    window.location.href = "/";
});

// Functions

function drawMyHand(handArray) {
    myHand.innerHTML = ""; 
    
    handArray.forEach(cardCode => {
        let img = document.createElement("img");
        img.src = `https://deckofcardsapi.com/static/img/${cardCode}.png`;
        img.className = "card-image clickable-card";
        img.dataset.code = cardCode; 
        
        img.addEventListener("click", () => {
            console.log("Attempting to play:", cardCode);
            socket.emit("crazyeights_play_card", {card_code: cardCode});
        });
        
        myHand.appendChild(img);
    });
}

function drawOpponentHand(position, cardCount) {
    const handContainer = document.getElementById(`hand-${position}`);

    handContainer.innerHTML = "";

    for (let i = 0; i < cardCount; i++) {
        let img = document.createElement("img");
        img.src = "https://deckofcardsapi.com/static/img/back.png"; 
        img.className = "card-image opponent-card"; 
        handContainer.appendChild(img);
    }
}

function updateSuitIcon(suitLetter) {
    if (suitLetter === 'H') suitIcon.innerText = "♥️";
    if (suitLetter === 'D') suitIcon.innerText = "♦️";
    if (suitLetter === 'C') suitIcon.innerText = "♣️";
    if (suitLetter === 'S') suitIcon.innerText = "♠️";
    if (suitLetter === "?") suitIcon.innerHTML = "❓";
    
    if (suitLetter === 'H' || suitLetter === 'D') {
        suitIcon.style.color = "#d32f2f";
    } else {
        suitIcon.style.color = "#212121";
    }
}

File: backend/sockets/crazyeights.py
"""Crazy Eights socket event handlers."""

from flask import request, session

from extensions import socketio
from registry import rooms, sid_player_obj_mapping
from sockets.validate import socket_validate


@socketio.on("crazyeights_player_join")
def crazyeights_player_join():
    if not socket_validate(session): return

    room_id = session.get("room")
    room = rooms.get(room_id)
    game = room.game
    username = session.get("username")
    print("DEBUG username:", username)  # add this
    print("DEBUG session:", dict(session))  # add this
    crazyeights_player = game.AddPlayer(request.sid, username)
    session["player_obj"] = crazyeights_player
    sid_player_obj_mapping[request.sid] = crazyeights_player 

    # Tell player the room size
    socketio.emit("crazyeights_room_size", {
        "room_size": room.num_players
    }, to=request.sid)

    # Tell everyone else in the room to render this new player
    socketio.emit("crazyeights_relay_player_info", {
        "id" : request.sid,
        "username": username,
        "hand": crazyeights_player.hand,
        "game_score": crazyeights_player.game_score
    }, to=room_id)

    # Tell this newly connected player about everyone who is already sitting at the table
    for player in game.players:
        socketio.emit("crazyeights_relay_player_info", {
            "id" : player.id,
            "username": player.username,
            "game_score" : player.game_score,
            "hand" : player.hand,
            "state" : player.state,
        }, to=request.sid)
        
    # If a round is already active, send them the current discard pile and active suit
    if game.discard_pile:
        socketio.emit("crazyeights_relay_board_info", {
            "top_card": game.discard_pile[-1]["code"],
            "current_suit": game.current_suit,
            "current_value": game.current_value
        }, to=request.sid)

@socketio.on("crazyeights_play_card")
def crazyeights_play_card(data):
    if not socket_validate(session): return
    
    room_id = session.get("room")
    room = rooms.get(room_id)
    sid = request.sid
    card_code = data.get("card_code")

    room.game.PlayCardRequest(sid, card_code)

@socketio.on("crazyeights_choose_suit")
def crazyeights_choose_suit(data):
    if not socket_validate(session): return

    room_id = session.get("room")
    room = rooms.get(room_id)
    sid = request.sid
    new_suit = data.get("suit")

    room.game.ChooseSuitRequest(sid, new_suit)

@socketio.on("crazyeights_leave_room")
def crazyeights_leave_room():
    if request.sid in sid_player_obj_mapping.keys():
        player_obj = sid_player_obj_mapping[request.sid]
        room_id = session.get("room")
        room = rooms.get(room_id)
        game = room.game

        if player_obj is not None: 
            player_obj.ClearHand()
            del sid_player_obj_mapping[request.sid]
            game.RemovePlayer(player_obj)
            room.player_count -= 1
            room.num_players -= 1
            print("Player left")
            socketio.emit("crazyeights_message", {"msg": f"{player_obj.username} left"}, to=room_id) 

            if len(game.players) == 1:
                last_player = game.players[0]

                socketio.emit("crazyeights_player_left", {
                    "id": last_player.id, 
                    "username":last_player.username
                }, to=room_id)
                
                game.RemovePlayer(last_player)
                del rooms[room_id]
                print("Room deleted")
            else:
                socketio.emit("crazyeights_player_left", {
                    "id": request.sid, 
                    "username": player_obj.username
                }, to=room_id)

